src/irmin-pack/unix/gc_worker.ml

```diff
@@ -267,57 +277,74 @@ module Make (Args : Gc_args.S) = struct
-        let pl : Payload.t = Fm.Control.payload (Fm.control fm) in
-        let end_offset =
-          Dispatcher.offset_of_suffix_poff dispatcher pl.suffix_end_poff
-        in
-        let len = Int63.Syntax.(end_offset - off) in
-        [%log.debug
-          "GC: transfer_loop iteration %d, offset %a, length %a"
-            (num_iterations - i + 1)
-            Int63.pp off Int63.pp len];
-        stats := Gc_stats.Worker.add_suffix_transfer !stats len;
-        let () = Dispatcher.read_bytes_exn dispatcher ~f:append_exn ~off ~len in
-        (* Check how many bytes are left, [4096*5] is selected because it is roughly the
-           number of bytes that requires a read from the block device on ext4 *)
-        if Int63.to_int len < 4096 * 5 then end_offset
-        else
-          let off = Int63.Syntax.(off + len) in
-          transfer_loop ~off (i - 1)
-    in
-    let new_end_suffix_offset =
-      transfer_loop ~off:new_prefix_end_offset num_iterations
+    (* Step 6. Calculate new suffix values. *)
+    let new_suffix, removable_chunk_idxs =
+      stats :=
+        Gc_stats.Worker.finish_current_step !stats
+          "suffix: calculate new values";
+      let suffix = Fm.suffix fm in
+      let soff = Dispatcher.soff_of_offset dispatcher new_suffix_start_offset in
+      (* Step 6.1. Calculate chunks that we have GCed. *)
+      let open struct
+        type chunk = { idx : int; end_suffix_off : int63 }
+      end in
+      let removable_chunks =
+        match Fm.Suffix.chunk_num suffix with
+        | 1 -> [] (* We never remove a single chunk. *)
+        | _ ->
+            Fm.Suffix.fold_chunks
+              (fun ~acc ~idx ~start_suffix_off ~end_suffix_off ~is_appendable ->
+                (* Remove chunks that end at or before our new split point.
+                   This will leave the chunk that starts with (or contains) the
+                   split point but remove all previous chunks.
+                   We never remove empty or appendable chunks. *)
+                let is_empty = start_suffix_off = end_suffix_off in
+                let ends_with_or_is_before_soff = end_suffix_off <= soff in
+                let is_removable =
+                  (not is_appendable)
+                  && (not is_empty)
+                  && ends_with_or_is_before_soff
+                in
+                if is_removable then { idx; end_suffix_off } :: acc else acc)
+              [] suffix
+      in
+      (* Step 6.2. Calculate the new chunk starting idx. *)
+      let chunk_start_idx =
+        match removable_chunks with
+        | [] -> Fm.Suffix.start_idx suffix
+        | last_removed_chunk :: _ -> succ last_removed_chunk.idx
+      in
+      (* Step 6.3. Calculate new dead bytes at the beginning of the suffix. *)
+      let suffix_dead_bytes =
+        match removable_chunks with
+        (* If no chunks are GCed, the dead bytes are equivalent to the physical
+           offset of new suffix offset. *)
```

Comment:
I'm wondering if this is true.
Suppose that we have a previous gc that left some dead bytes in the first chunk (prev_dead_bytes), and then a second gc that does not remove any chunks.  Shouldn't the new_dead_bytes be the prev_dead_bytes? Am I missing something?

Reply:
Thank you for questioning this assertion. :smile: 

I do think it is true. 

Lets start with a suffix after previous GC w/ dead bytes:
```
[          |                                                       ][               ]
0      dead bytes                                                                   N
```

Then do a new GC that does not remove a chunk:
```
                                         new suffix offset poff                      
[          |                                        |              ][               ]
0    prev dead bytes                                                                N
```

Does this visual help show the logic of the comment? If it were set to prev dead bytes, we would still read from the suffix in (prev dead bytes, new suffix offset poff) instead of the prefix, which is what we want.

Reply:
Hm, you are right. I'm still a bit uncertain, but I can't yet pinpoint where my issue is. So no need to block this, I'll add a test later on if I find a problem. Thanks for the explanation!